File: dotnet/ChargeBee/Api/ApiException.cs
using System;
using System.Net;
using System.Runtime.Serialization;
using System.Collections.Generic;

namespace ChargeBee.Api
{
    public class ApiException : ApplicationException
    {

		private string ErrorType = "";
		private string ErrorParam = "";

		public ApiException (HttpStatusCode httpStatusCode, Dictionary<string, string> errorResp)
			: base (errorResp ["message"])
        {
			this.HttpStatusCode = httpStatusCode;
			errorResp.TryGetValue ("type", out ErrorType);
			this.Code = errorResp ["api_error_code"];

			//Deprecated fields.
			this.ApiCode = errorResp ["error_code"];
			this.ApiMessage = errorResp ["error_msg"];

			errorResp.TryGetValue("param", out ErrorParam);
        }

        public HttpStatusCode HttpStatusCode { get; set; }

		public string Type { 
			get {
				return this.ErrorType;
			}
		}

        public string Code { get; set; }

        public string Param { 
			get {
				return this.ErrorParam;
			}
		}

        [System.Obsolete("Use HttpStatusCode")]
        public HttpStatusCode HttpCode { 
			get {
				return HttpStatusCode;
			}
		}

        [System.Obsolete("Use Code")]
		public string ApiCode { get; set; }

        [System.Obsolete("Use Param")]
		public string Parameter { 
			get { 
				return Param;
			} 
		}

        [System.Obsolete("Use Messsge")]
        public string ApiMessage { get; set; }

    }
}

File: java/src/main/java/com/chargebee/APIException.java
package com.chargebee;

import org.json.*;

public class APIException extends RuntimeException {

    public final JSONObject jsonObj;
    
    /**
     * Use httpStatusCode instead.
     * @deprecated
     */
    @Deprecated
    public final int httpCode;
    
    public final int httpStatusCode;
    public final String type;
    public final String code;
    public final String param;

    /**
     * Use getMessage instead.
     * @deprecated
     */
    @Deprecated
    public final String message;
    

    public APIException(int httpStatusCode,JSONObject jsonObj) throws Exception {
        super(jsonObj.getString("message"));
        this.jsonObj = jsonObj;
        this.httpCode = httpStatusCode;
        this.httpStatusCode = httpStatusCode;
        this.code = jsonObj.getString("code");
        this.type = jsonObj.optString("type");
        this.param = jsonObj.optString("param");

        this.message = jsonObj.getString("error_msg");
    }
    
    
    public boolean isParamErr() {
        return param != null;
    }

    @Override
    public String toString() {
        try {
            return jsonObj.toString(2);
        } catch (JSONException ex) {
            throw new RuntimeException(ex);
        }
    }
}
